File: python/llm/providers/gemini.py
"""Google Gemini Provider。

调用 Gemini generateContent HTTP API，映射统一 ChatRequest/Response。"""

from __future__ import annotations

import json
from urllib.error import HTTPError, URLError
from urllib.parse import urlencode
from urllib.request import Request, urlopen

from config.models import ProviderSettings
from llm.models import ChatRequest, ChatResponse, LlmError

DEFAULT_BASE_URL = "https://generativelanguage.googleapis.com"


class GeminiProvider:
    def __init__(self, settings: ProviderSettings) -> None:
        self.settings = settings

    @property
    def kind(self) -> str:
        return "gemini"

    @property
    def supports_tools(self) -> bool:
        return False

    def _endpoint(self, model: str) -> str:
        base = self.settings.base_url.strip().rstrip("/") or DEFAULT_BASE_URL
        return f"{base}/v1beta/models/{model}:generateContent"

    def complete(self, request: ChatRequest) -> ChatResponse:
        model = request.model or self.settings.model
        system = "\n".join(m.content for m in request.messages if m.role == "system")
        contents = [
            {
                "role": "model" if m.role == "assistant" else "user",
                "parts": [{"text": m.content}],
            }
            for m in request.messages
            if m.role != "system"
        ]
        payload: dict[str, object] = {
            "contents": contents,
            "generationConfig": {
                "temperature": request.temperature,
                "maxOutputTokens": request.max_tokens,
            },
        }
        if system:
            payload["systemInstruction"] = {"parts": [{"text": system}]}

        query = urlencode({"key": self.settings.api_key})
        url = f"{self._endpoint(model)}?{query}"
        body = json.dumps(payload).encode("utf-8")
        req = Request(
            url,
            data=body,
            headers={"Content-Type": "application/json"},
            method="POST",
        )
        try:
            with urlopen(req, timeout=120) as response:  # noqa: S310
                data = json.loads(response.read().decode("utf-8"))
        except HTTPError as error:
            raise LlmError(f"gemini http {error.code}") from error
        except URLError as error:
            raise LlmError(f"gemini transport: {error}") from error

        candidates = data.get("candidates") or []
        if not candidates:
            raise LlmError("empty response")
        parts = candidates[0].get("content", {}).get("parts") or []
        reply = ""
        for part in parts:
            if part.get("text"):
                reply = str(part["text"]).strip()
                break
        if not reply:
            raise LlmError("empty response")
        finish = candidates[0].get("finishReason")
        return ChatResponse(reply=reply, finish_reason=str(finish) if finish else None)

File: crates/agent/src/model/gemini.rs
//! Google Gemini provider — reqwest 薄封装。

use super::{ChatRequest, ChatResponse, LlmError, LlmProvider, ProviderSettings};
use async_trait::async_trait;
use serde_json::json;

const DEFAULT_BASE_URL: &str = "https://generativelanguage.googleapis.com";

/// Gemini provider。
pub struct GeminiProvider {
    settings: ProviderSettings,
}

impl GeminiProvider {
    pub fn new(settings: ProviderSettings) -> Self {
        Self { settings }
    }

    fn endpoint(&self, model: &str) -> String {
        let base = self.settings.base_url.trim().trim_end_matches('/');
        let base = if base.is_empty() {
            DEFAULT_BASE_URL
        } else {
            base
        };
        format!("{base}/v1beta/models/{model}:generateContent")
    }
}

#[async_trait]
impl LlmProvider for GeminiProvider {
    fn kind(&self) -> &'static str {
        "gemini"
    }

    async fn complete(&self, request: &ChatRequest) -> Result<ChatResponse, LlmError> {
        // Gemini 协议：system 走 systemInstruction；user/assistant 平铺为 contents。
        let system = request
            .messages
            .iter()
            .filter(|msg| msg.role == "system")
            .map(|msg| msg.content.clone())
            .collect::<Vec<_>>()
            .join("\n");
        let contents: Vec<_> = request
            .messages
            .iter()
            .filter(|msg| msg.role != "system")
            .map(|msg| {
                json!({
                    "role": if msg.role == "assistant" { "model" } else { "user" },
                    "parts": [{"text": msg.content}]
                })
            })
            .collect();

        let mut payload = json!({
            "contents": contents,
            "generationConfig": {
                "temperature": request.temperature,
                "maxOutputTokens": request.max_tokens,
            }
        });
        if !system.is_empty() {
            payload["systemInstruction"] = json!({ "parts": [{ "text": system }] });
        }

        let client = reqwest::Client::new();
        let response = client
            .post(self.endpoint(&request.model))
            .query(&[("key", &self.settings.api_key)])
            .json(&payload)
            .send()
            .await
            .map_err(|error| LlmError::Transport(format!("gemini: {error}")))?;
        let body: serde_json::Value = response
            .json()
            .await
            .map_err(|error| LlmError::Transport(format!("gemini parse: {error}")))?;

        let reply = body
            .pointer("/candidates/0/content/parts/0/text")
            .and_then(serde_json::Value::as_str)
            .map(|text| text.trim().to_string())
            .filter(|text| !text.is_empty())
            .ok_or(LlmError::EmptyResponse)?;

        let finish_reason = body
            .pointer("/candidates/0/finishReason")
            .and_then(serde_json::Value::as_str)
            .map(|s| s.to_string());

        Ok(ChatResponse {
            reply,
            finish_reason,
            tool_calls: Vec::new(),
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn endpoint_uses_model_and_key() {
        let provider = GeminiProvider::new(ProviderSettings {
            provider_type: "gemini".into(),
            api_key: String::from("k"),
            base_url: "".into(),
            model: "gemini-1.5-pro".into(),
        });
        assert_eq!(
            provider.endpoint("gemini-1.5-pro"),
            "https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-pro:generateContent"
        );
    }
}
